Fix regression with inout in enum and subscript signatures

slavapestov opened the pull request:
We were overzealously banning 'inout' in enum and subscript signatures. It can actually appear in the argument list of a nested function type.

Fixes <https://bugs.swift.org/browse/SR-7890>.

slavapestov committed:
Remove duplicate diagnostic between Parse and Sema

slavapestov committed:
Sema: Remove dead code from IUO removal

slavapestov committed:
Sema: Clear more type resolution flags for single-argument function types

This fixes a 4.2 regression where enums and subscripts could not
contain single-argument function types with an 'inout' parameter,
because we erroneously diagnosed the 'inout' as if it appeared
at the top level of the enum case or subscript index type.

Fixes <https://bugs.swift.org/browse/SR-7890>.

slavapestov commented:
@swift-ci Please smoke test

xedin reviewed (commented):
Thanks for looking into this @slavapestov ! It looks like it might be more complicated to fix because now it rejects single argument with inout and no parens...

slavapestov commented:
That was just an overzealous cleanup. I can drop one of the patches here.

slavapestov committed:
Fix a couple of nits

slavapestov committed:
Parse: Fix crash in error recovery path with specifier before parameter name

We should let Sema set ParamDecl's specifier always, instead of setting
it in the parser in this odd corner case where we're recovering from
'inout' being written before the parameter name.

slavapestov commented:
@swift-ci Please smoke test
